DRIVERS-2672 Add OIDC machine workflow spec. (#1471)
mongodb/specifications: source/auth/tests/mongodb-oidc.rst

  AWS provider logic.
- Start 10 threads and run 100 ``find`` operations in each thread that all
  succeed.
- Assert that the callback was called 1 time.
- Close the client.

(2) OIDC Callback Validation
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**2.1 Valid Callback Inputs**

- Create a ``MongoClient`` configured with an OIDC callback that validates its
  inputs and returns a valid access token.
- Perform a ``find`` operation that succeeds.
- Assert that the OIDC callback was called with the appropriate inputs,
  including the timeout parameter if possible.
- Close the client.

**2.2 OIDC Callback Returns Null**

- Create a ``MongoClient`` configured with an OIDC callback that returns
  ``null``.
- Perform a ``find`` operation that fails.
- Close the client.

**2.3 OIDC Callback Returns Missing Data**

- Create a ``MongoClient`` configured with an OIDC callback that returns data
  not conforming to the ``OIDCCredential`` with missing fields.
- Perform a ``find`` operation that fails.
- Close the client.

**2.4 Invalid Client Configuration with Callback**

- Create a ``MongoClient`` configured with an OIDC callback and auth mechanism
  property ``PROVIDER_NAME:aws``.

> - Assert it returns a client configuration error.

Reviewer: This prose test is already covered by the unified connection string tests above. Isn't this an unnecessary duplicate?

Author: The updated Auth connection string tests don't support passing an OIDC callback, so I believe this prose test covers a case that the Auth connection string tests can't. The previous Auth connection string OIDC tests did include a "callback" param, but it was not documented in the test format README, was only implemented in 1 or 2 drivers, and was confusing with the updated OIDC callback API, so I decided to replace it with a prose test instead.

Drivers MUST test the following scenarios:

- ``Callback-Driven Auth``
- ``Callback Validation``
- ``Speculative Authentication``
- ``Reauthentication``
- ``Separate Connections Avoid Extra Callback Calls``

.. sectnum::

> Drivers MUST be able to authenticate against a server configured with either one or two configured identity providers.

Reviewer 2: Is it enough to test with either one (pick one), or should it succeed with either (both)?

> Drivers MUST be able to authenticate against a server configured with either one or two configured identity providers.

Reviewer 2: ```suggestion
In the tests below, drivers will authenticate against a server configured with one identity provider (``MONGODB_URI_SINGLE``), and against a server configured with two identity providers (``MONGODB_URI_MULTI``).
```

Author: Rewrote this section to clarify what drivers should test.

Note that typically the preconfigured Atlas Dev clusters are used for testing, in Evergreen and localy.  The URIs can be fetched
from the ``drivers/oidc`` Secrets vault, see `vault instructions`_.  Use ``OIDC_ATLAS_URI_SINGLE`` for ``MONGODB_URI_SINGLE`` and

> ``OIDC_ATLAS_URI_MULTI`` for ``OIDC_ATLAS_URI_MULTI``.

Reviewer 3: Should this be `MONGODB_URI_MULTI` as the tests use below?

> ``OIDC_ATLAS_URI_MULTI`` for ``OIDC_ATLAS_URI_MULTI``.

Author: Fixed!

If using local servers is preferred, using the `Local Testing`_ method,
use ``mongodb://localhost/?authMechanism=MONGODB-OIDC`` for ``MONGODB_URI_SINGLE`` and
``mongodb://localhost:27018/?authMechanism=MONGODB-OIDC&directConnection=true&readPreference=secondaryPreferred``
for ``MONGODB_URI_MULTI`` because the other server is a secondary on a replica set, on port ``27018``.

> The default OIDC client used in the tests will be configured with ``MONGODB_URI_SINGLE`` and a valid request callback handler

Reviewer 3: Should this be implemented with `CUSTOM_TOKEN_CALLBACK` and `PROVIDER_NAME=custom`?

Reviewer 4: No, it will depend on where it is being run.  We'll want to use the same unified tests on Azure, AWS, and GCP, hence the placeholder.

Single Principal Implicit Username
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> - Create default OIDC client.

Reviewer 3: ```suggestion
- Create default OIDC client with `authMechanism=MONGODB-OIDC`.
```

- Perform a ``find`` operation. that succeeds.
- Close the client.

Single Principal Explicit Username
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> - Create a client with ``MONGODB_URI_SINGLE``, a username of ``test_user1``, and the OIDC request callback.

Reviewer 3: ```suggestion
- Create a client with ``MONGODB_URI_SINGLE``, a username of ``test_user1``, `authMechanism=MONGODB-OIDC`, and the OIDC request callback.
```

- Perform a ``find`` operation that succeeds.
- Close the client.

Multiple Principal User 1
~~~~~~~~~~~~~~~~~~~~~~~~~

> - Create a client with ``MONGODB_URI_MULTI``, a username of ``test_user1``, and the OIDC request callback.

Reviewer 3: ```suggestion
- Create a client with ``MONGODB_URI_MULTI``, a username of ``test_user1``, `authMechanism=MONGODB-OIDC`, and the OIDC request callback.
```

- Perform a ``find`` operation that succeeds.
- Close the client.

Multiple Principal User 2
~~~~~~~~~~~~~~~~~~~~~~~~~
- Create a request callback that reads in the generated ``test_user2`` token file.

> - Create a client with ``MONGODB_URI_MULTI``, a username of ``test_user2``, and the OIDC request callback.

Reviewer 3: ```suggestion
- Create a client with ``MONGODB_URI_MULTI``, a username of ``test_user2``, `authMechanism=MONGODB-OIDC`, and the OIDC request callback.
```

- Perform a ``find`` operation that succeeds.
- Close the client.

Multiple Principal No User
~~~~~~~~~~~~~~~~~~~~~~~~~~

> - Create a client with ``MONGODB_URI_MULTI``, no username, and the OIDC request callback.

Reviewer 3: ```suggestion
- Create a client with ``MONGODB_URI_MULTI``, no username, `authMechanism=MONGODB-OIDC`, and the OIDC request callback.
```

- Assert that a ``find`` operation fails.
- Close the client.

Allowed Hosts Blocked
~~~~~~~~~~~~~~~~~~~~~

> - Create a default OIDC client, with an ``ALLOWED_HOSTS`` that is an empty list.

Reviewer 2: Could we either split this into two files, or into two distinct sections, one for machine workflow tests, and the other for human workflow tests (which can all be omitted, for drivers that do not implement them)?

Author: Done!

- Close the client.
- Create a client that uses the url ``mongodb://localhost/?authMechanism=MONGODB-OIDC&ignored=example.com`` a request callback, and an
  ``ALLOWED_HOSTS`` that contains ["example.com"].
  ``ALLOWED_HOSTS`` that contains ``["example.com"]``.
- Assert that a ``find`` operation fails with a client-side error.
- Close the client.

Lock Avoids Extra Callback Calls
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- Clear the cache.
- Create a request callback that returns a token that will expire soon, and
  a refresh callback.  Ensure that the request callback has a time delay, and
  that we can record the number of times each callback is called.
- Spawn two threads or async operations that do the following:
  - Create a client with the callbacks.
  - Run a find operation that succeeds.
  - Close the client.
  - Create a new client with the callbacks.
  - Run a find operation that succeeds.
  - Close the client.
- Join the two threads or simultaneously call the async operations
- Ensure that the request callback has been called once, and the refresh
  callback has been called twice, or that no async function has been
  entered simultaneously.

> AWS Automatic Auth

Reviewer 3: Shouldn't we be keeping these tests and changing the URIs from localhost to `MONGODB_URI_SINGLE` and `MONGODB_URI_MULTI` with a `PROVIDER_NAME` of aws?

Reviewer 4: `MONGODB_URI_MULTI` will only be used for human workflows, which will continue to be prose tests only.

  handshake, and again during reauthentication).
- Close the client.

3. Authentication failures with cached tokens retry with a new token
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- Create a ``MongoClient`` configured with ``retryReads=false`` and a custom
  OIDC callback that implements the AWS provider logic.
- Set a fail point for ``find`` commands of the form:

.. code:: javascript

    {
      configureFailPoint: "failCommand",
      mode: {
        times: 1
      },
      data: {
        failCommands: [
          "find"
        ],
        closeConnection: true
      }
    }

- Perform a ``find`` operation that fails. This is to force the ``MongoClient``
  to cache an access token.

> - Set a fail point for ``saslStart`` commands of the form:

Reviewer 4: This doesn't raise a failure because it will successfully connect in speculative auth.

Author: Replaced this test with one that works with speculative auth.

.. code:: javascript

    {
      configureFailPoint: "failCommand",
      mode: {
        times: 2
      },
      data: {
        failCommands: [
          "saslStart"
        ],
        errorCode: 18
      }
    }

- Perform a ``find`` operation that fails.
- Verify that the callback was called 2 times during connection handshake (once
  to get the initial token, and once to refresh the token after the
  authentication failure).
- Close the client.

4. Reauthentication messages are sent
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- Create a ``MongoClient`` configured with a custom OIDC callback that
  implements the AWS provider logic.
- Perform a ``find`` operation that succeeds.
- Set fail points for ``find`` and ``saslStart`` of the form:

.. code:: javascript

    {
      configureFailPoint: "failCommand",
      mode: {
        times: 1
      },
      data: {
        failCommands: [
          "find"
        ],
        errorCode: 391
      }
    }

    {
      configureFailPoint: "failCommand",
      mode: {
        times: 2
      },
      data: {
        failCommands: [
          "saslStart"
        ],
        errorCode: 18
      }
    }

> - Perform a ``find`` operation that fails.

Reviewer 4: Same, this will succeed on spec auth.

Author: Replaced this test with one that works with speculative auth.
